lib/galaxy/model/util.py

```diff
@@ -3,31 +3,94 @@
+    ids_dict: stores the state of dataset and its associated hda and history, key=str(dataset_id), value could be one of following.
+        -1: dataset deleted
+        0: dataset not deleted, history deleted, hda deleted
+        1: dataset not deleted, history deleted, hda not deleted
+        2: dataset not deleted, history not deleted, hda deleted
+        3: dataset not deleted, history not deleted, hda not deleted
+
+    Return:
+    disk usage counted toward deleted disk usage for the input dataset
+    """
+    id_key = str(dataset_id)
+    # already counted, state fixed
+    if id_key in ids_dict and (ids_dict[id_key] == -1 or ids_dict[id_key] == 3):
+        return 0
+    if id_key not in ids_dict:       # first count; count all possibles
+        if dataset_deleted:
+            ids_dict[id_key] = -1    # must count
+            return dataset_disk_usage
+        elif hda_deleted:
+            if history_deleted:      # possible
+                ids_dict[id_key] = 0
+                return dataset_disk_usage
+            else:                    # possible
+                ids_dict[id_key] = 2
+                return dataset_disk_usage
+        else:
+            if history_deleted:     # possible
+                ids_dict[id_key] = 1
+                return dataset_disk_usage
+            else:                   # impossible
+                ids_dict[id_key] = 3
+                return 0
+    else:                           # repeat count
+        if hda_deleted:
+            if history_deleted:
+                ids_dict[id_key] = ids_dict[id_key] | 0
+            else:
+                ids_dict[id_key] = ids_dict[id_key] | 2
+        else:
+            if history_deleted:
+                ids_dict[id_key] = ids_dict[id_key] | 1
+            else:
+                ids_dict[id_key] = 3
+        if ids_dict[id_key] == 3:   # remove mis-count
+            return -dataset_disk_usage
+        else:
+            return 0
+
+
 def pgcalc(sa_session, id, dryrun=False):
     """
     Utility method for quickly recalculating user disk usage in postgres.
 
     TODO: Check against the recently updated versions of sqlalchemy if this
     'special' postgresql version is even necessary.
     """
-    sql_calc = """SELECT COALESCE(SUM(total_size), 0)
-                  FROM (  SELECT DISTINCT ON (d.id) d.total_size, d.id
-                          FROM history_dataset_association hda
-                               JOIN history h ON h.id = hda.history_id
-                               JOIN dataset d ON hda.dataset_id = d.id
-                          WHERE h.user_id = :id
-                                AND h.purged = false
+    ids_dict = {}
+    total = 0
+    deleted = 0
+    query_history = """SELECT id, deleted
+                    FROM history
+                        WHERE user_id = :id
+                            AND purged = false;"""
+    for history_id, history_deleted in sa_session.execute(query_history, {'id': id}).fetchall():
```

Comment:
Did you try doing this in a single query, this and the inner `query_hda_dataset`?  Was it similar in performance?

Reply:
I just followed the same way used in `model.User.calculate_disk_usage`. Someone logged this, '# this can be a huge number and can run out of memory, so we avoid the mappers'.

Reply:
Sounds good to me.

Reply:
Thanks.